cdn/single-origin.yml

```diff
@@ -113,58 +115,74 @@ Metadata:
         default: AWS WAF web ACL ID
       AcmCertificateArn:
         default: Certificate ARN
+      CertificateAttachment:
+        default: Certificate attachment
 Parameters:
   ProjectTag:
     Type: String
+    AllowedPattern: ^.+$
+    ConstraintDescription: must not be blank
     Description: >
       The value used for the Project tag on resources that support tagging.
   Cnames:
     Type: CommaDelimitedList
     Description: >
-      The CNAMEs (alternate domain names) for the distribution, as a
+      (optional) The CNAMEs (alternate domain names) for the distribution, as a
       comma-delimited list (e.g. "cdn1.example.com,cdn2.example.com")
   OriginDomain:
     Type: String
+    AllowedPattern: ^.+\.[a-zA-Z]+$
+    ConstraintDescription: >
+      must be a hostname or of the format mybucket.s3.amazonaws.com
     Description: >
       The DNS name of the Amazon Simple Storage Service (S3) bucket or the HTTP
       server from which you want CloudFront to get objects for this origin.
+      (e.g. "example.com" or "mybucket.s3.amazonaws.com")
   OriginPath:
     Type: String
+    AllowedPattern: ^$|^\/..+(?<!\/)$
```

Comment:
Does this look right?

Reply:
Hrmm... in a normal language I would have written that as `/^\/.+[^\/]$` (starts with `/`, followed by 0 or more of anything, ending with not-`/`).  Do those `AllowedPattern`s have a funny format or something?

Reply:
I'm just bad at regular expressions. Your way seems better. I think I had the extra dot in there to make sure it contained a dot.